developer_manual/client_apis/OCS/ocs-openapi.rst

```diff
@@ -0,0 +1,731 @@
+They hide problems that your code has.
+Especially the later one is dangerous because it will never fail when any problem exists in your code.
+They should only be used to speed up the initial porting process.
+
+Best practices
+--------------
+
+Note that you can find a step-by-step tutorial after this section.
+You can also read the tutorial before reading the best practices.
+
+PREFER to expose your APIs using OCS
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+It provides a more standardized and easier way to write your APIs.
+Other methods are considered legacy.
+For details take a look at :ref:`OCS <ocscontroller>`.
+
+.. collapse:: Examples
+
+    .. code-block:: php
+        :caption: Bad
+
+        class SomeController extends ApiController {
+            ...
+
+            public function someControllerMethod(): JSONResponse {
+                ...
+                return JSONResponse(...);
+            }
+        }
+
+    .. code-block:: php
+        :caption: Good
+
+        class SomeController extends OCSController {
+            ...
+
+            public function someControllerMethod(): DataResponse {
+                ...
+                return DataResponse(...);
+            }
+        }
+
+DO type controller and helper methods as explicit as possible
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The closer you narrow down a type without violating any constraints the better the resulting specification will be.
+Psalm will catch these problems for you if you configured the issue handlers mentioned above correctly.
+
+.. collapse:: Examples
+
+    .. code-block:: php
+        :caption: Bad
+
+        /**
+         * @return array
+         */
+        public function someHelperMethod() {
+            ...
+            return [
+                "id" => id,
+                "name" => name,
+            ];
+        }
+
+    .. code-block:: php
+        :caption: Good
+
+        /**
+         * @return array{id: int, name: string}
+         */
+        public function someHelperMethod() {
+            ...
+            return [
+                "id" => id,
+                "name" => name,
+            ];
+        }
+
+PREFER to use `null` to represent empty data
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Your API should be designed in a way that represents empty data with `null`.
+
+There is a problem with PHP and arrays that get converted to JSON.
+JSON has lists and objects while PHP only has arrays.
+If you were to return an empty array in PHP it will always turn into `[]` in JSON.
+This is not a problem for endpoints that always return lists, but most endpoints return a single JSON object.
+For those endpoints returning `[]` in PHP is a problem because the consumer will either get `[]` or `{...}` which is hard to handle.
+
+If you are not able to use `null` for whatever reason, use `new \stdClass()` instead.
+It will get correctly converted into `{}` in the JSON response.
```

Comment:
But it breaks the XML response, making your OCS not fully OCS compatible.
See https://github.com/nextcloud/spreed/blob/c4a0f78e478fdca0353d3c47a6fcaaa0a0b30711/lib/Model/Message.php#L162-L167 + https://github.com/nextcloud/spreed/blob/8225b4443b5678c1034757d6b5a989054e4377f4/lib/Controller/AEnvironmentAwareController.php#L64-L83 for a painful learned work around

Reply:
It doesn't anymore since https://github.com/nextcloud/server/pull/38745

Reply:
Should then mention 28+ as requirement and the change should make it into the dev-docs:
https://github.com/nextcloud/documentation/blob/master/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_28.rst#changed-apis-1

That being said, let's hope no one uses a non-empty stdClass:
https://www.php.net/manual/en/class.stdclass.php#stdclass.examples

Reply:
28+ requirement is mentioned at the beginning of the document

Reply:
Well there is a difference between "Supports at least 28+" and "Requires at least 28+"

Just add a simple comment:
```suggestion
If you are not able to use `null` for whatever reason, use `new \stdClass()` instead.
It will get correctly converted into `{}` in the JSON response on Nextcloud 28 and later.
```